IF-Match/IF-NOT-MATCH header support for VMSS creation/update

grizzlytheodore opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
Workitem: https://msazure.visualstudio.com/One/_workitems/edit/24979186
design: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/1372

This looks like a complicated PR; it is because VM/VMSS creation in simple parameter set uses a legacy [method ](https://github.com/Azure/azure-powershell/blob/main/src/Compute/Compute/Manual/VirtualMachineScaleSetCreateOrUpdateMethod.cs#L674)which makes it hard to pass in headers. 

So I implemented the same design that Dingmeng and I came up with when I had to add custom header auxiliary token for 'cross tenant sharing' feature. PR reference: https://github.com/Azure/azure-powershell/pull/17609

<!-- Please add a brief description of the changes made in this PR. If you have an ongoing or finished cmdlet design, please paste the link below. -->

## Mandatory Checklist

- Please choose the target release of Azure PowerShell
  - [x] [General release](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Public preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Private preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Engineering build](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] N/A

- [x] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For SDK-based development mode, update `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`.
        * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. 
    * For autorest-based development mode, include the changelog in the PR description.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** adjust version of module manually in pull request

grizzlytheodore commented:
![image](https://github.com/Azure/azure-powershell/assets/10857691/cdea7a86-6329-4636-8abe-23aa1489b784)
![image](https://github.com/Azure/azure-powershell/assets/10857691/eca73f09-4327-49e3-9599-a654d62e74ec)

grizzlytheodore commented on src/Compute/Compute/Manual/VirtualMachineScaleSetCreateOrUpdateMethod.cs:
```diff
@@ -561,6 +563,25 @@ private async Task<ResourceConfig<VirtualMachineScaleSet>> SimpleParameterSetOrc
+                        }
+                    }
+                }
+
```
aux header was not being supported for flex mode simple parameter set creation. Fixing it here

grizzlytheodore replied:
thought this is applicable since Flex mode can be now created with VM Profile or without

Sandido replied:
Is this code still needed? @grizzlytheodore

grizzlytheodore replied:
yup. still needed since auxHeader can be passed in either way - flex mode or uniform mode

Nickcandy commented:
Could you add some tests for the new parameters?

Nickcandy commented:
@grizzlytheodore Just a friendly reminder: our code freeze deadline is at 6 PM PST on April 22nd. If your code is intended for this release, please ensure it is completed as soon as possible.

grizzlytheodore commented:
> Could you add some tests for the new parameters?

hello, I wrote why I wasn't able to write a test for this header feature. https://github.com/Azure/azure-powershell/pull/24657/#issuecomment-2065474397

grizzlytheodore commented:
> @grizzlytheodore Just a friendly reminder: our code freeze deadline is at 6 PM PST on April 22nd. If your code is intended for this release, please ensure it is completed as soon as possible.

We can catch the next release @nickcandy

YanaXu commented:
/azp run

YanaXu commented:
Hi @grizzlytheodore, since no tests are added and the plain text updates are a lot, could you add one of your team member to review this PR too?
And the update is too much because you update the format of csharp files. If it's OK, could you slipt the format part and feature part in 2 PRs? It'll be easier for us to review it.

grizzlytheodore commented on src/Compute/Compute/Strategies/ComputeRp/VirtualMachineScaleSetStrategy.cs:
```diff
@@ -220,98 +226,112 @@ public static class VirtualMachineScaleSetStrategy
                 resourceGroup: resourceGroup,
                 name: name,
-                createModel: engine => new VirtualMachineScaleSet()
-                {
```
easier to view changes in split view

grizzlytheodore commented:
> Hi @grizzlytheodore, since no tests are added and the plain text updates are a lot, could you add one of your team member to review this PR too? And the update is too much because you update the format of csharp files. If it's OK, could you slipt the format part and feature part in 2 PRs? It'll be easier for us to review it.

I've removed the format changes. they came unintentionally when using a tool but I've removed it. 

I will get my teammates to review it thank you. @YanaXu

grizzlytheodore commented on src/Compute/Compute/Custom/VirtualMachineScaleSetsOperationsExtensions.cs:
```diff
@@ -13,12 +13,16 @@ public static partial class VirtualMachineScaleSetsOperationsExtensions
-                return operations.CreateOrUpdate(resourceGroupName, vmssName, virtualMachineScaleSet);
+                return operations.CreateOrUpdate(resourceGroupName, vmssName, virtualMachineScaleSet, ifMatch, ifNoneMatch);
             }
             virtualMachineScaleSet.RemoveAuxAuthHeader();
```
dettaching header before calling function. to check

grizzlytheodore replied:
saved on auxAuthHeader variable so this is right approach. resolved

Sandido commented on src/Compute/Compute/Custom/VirtualMachinesOperationsExtensions.cs:
```diff
@@ -13,12 +13,16 @@ public static partial class VirtualMachinesOperationsExtensions
             var auxAuthHeader = virtualMachine.GetAuxAuthHeader();
+            var ifMatchResult = virtualMachine.GetIfMatchIfNoneMatch();
+            var ifMatch = ifMatchResult.Item1;
+            var ifNoneMatch = ifMatchResult.Item2;
```
Is there a better way to retrieve the ifMatch and ifNoneMAtch? Maybe instead of making it a tuple, make it a class that has better named properties? OR can the name of the properties in the tuple be made to be more descriptive?

grizzlytheodore replied:
i could retrieve them individually with seprate functions, but I just wanted to make it one function so I went with a Tuple. 
Item1, Item2 in this case are just getter methods for tuples, not named properties

Sandido commented on src/Compute/Compute/Generated/VirtualMachineScaleSet/VirtualMachineScaleSetCreateOrUpdateMethod.cs:
```diff
@@ -176,14 +176,14 @@ public override void ExecuteCmdlet()
                                         resourceGroupName,
                                         vmScaleSetName,
-                                        parameters,null,null,
+                                        parameters, this.IfMatch, this.IfNoneMatch,
```
Make sure to run tests where these are null to ensure this does not throw a null ref exception.

grizzlytheodore replied:
here is running with ifMatch and ifNoneMatch not provided: 
![image](https://github.com/Azure/azure-powershell/assets/10857691/798af67a-501c-44ae-8301-ff2e9e2a586c)
![image](https://github.com/Azure/azure-powershell/assets/10857691/513166bc-93a6-40f0-b89a-839192c69de3)
![image](https://github.com/Azure/azure-powershell/assets/10857691/94e34c63-40d0-4acc-976b-0dafb37b0434)

Sandido commented on src/Compute/Compute/Generated/VirtualMachineScaleSet/VirtualMachineScaleSetUpdateMethod.cs:
```diff
@@ -81,8 +81,8 @@ public override void ExecuteCmdlet()
                     var result = (this.VirtualMachineScaleSetUpdate == null)
-                                 ? VirtualMachineScaleSetsClient.CreateOrUpdate(resourceGroupName, vmScaleSetName, parameters)
-                                 : VirtualMachineScaleSetsClient.Update(resourceGroupName, vmScaleSetName, parametersupdate);
+                                 ? VirtualMachineScaleSetsClient.CreateOrUpdate(resourceGroupName, vmScaleSetName, parameters, this.IfMatch, this.IfNoneMatch)
```
same concern about testing for null ref exception.

grizzlytheodore replied:
resolving with screenshots above

grizzlytheodore commented:
to show it working without adding ifMatch/IfNoneMatch

Sandido commented on src/Compute/Compute/Manual/VirtualMachineScaleSetCreateOrUpdateMethod.cs:
```diff
@@ -561,6 +563,25 @@ private async Task<ResourceConfig<VirtualMachineScaleSet>> SimpleParameterSetOrc
                 var hostGroup = resourceGroup.CreateDedicatedHostGroupSubResourceFunc(_cmdlet.HostGroupId);
 
+                Dictionary<string, List<string>> auxAuthHeader = null;
+                if (!string.IsNullOrEmpty(_cmdlet.ImageReferenceId))
```
should you only fill the auxheader when ImageReferenceId is passed? Should the auxheader be filled from the user in any other case?

grizzlytheodore replied:
I remember now. Currently users cannot pass in aux header in PS.  We are auto populating aux header with correct token when user references a gallery image that is cross tenant. so currently in powershell aux header is only used in this specific case

Sandido commented on src/Compute/Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs:
```diff
@@ -182,6 +182,18 @@ public class UpdateAzureVMCommand : VirtualMachineBaseCmdlet
+        [Parameter(
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The ETag of the transformation. Omit this value to always overwrite the current resource. Specify the last-seen ETag value to prevent accidentally overwriting concurrent changes.")]
```
Is this the etag or ifmatch? HelpMessage needs to be more clear.

grizzlytheodore replied:
this is ifMatch from swagger

Sandido replied:
@grizzlytheodore , even so, I think it needs to improved, or we need to understand it better bc I am still confused by this. Maybe the swagger description needs to be updated too.

grizzlytheodore replied:
@Sandido  improved, as requested !

Sandido commented on src/Compute/Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs:
```diff
@@ -182,6 +182,18 @@ public class UpdateAzureVMCommand : VirtualMachineBaseCmdlet
+        [Parameter(
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "Set to '*' to allow a new record set to be created, but to prevent updating an existing record set. Other values will result in error from server as they are not supported.")]
```
What does this mean? Needs to be more clear.

grizzlytheodore replied:
got it from swagger as well

grizzlytheodore replied:
improved

haagha reviewed (dismissed):
Comments to check a few things and behaviour when header is not specified.

grizzlytheodore commented on src/Compute/Compute/Generated/VirtualMachineScaleSet/VirtualMachineScaleSetCreateOrUpdateMethod.cs:
```diff
@@ -401,5 +401,16 @@ private int convertAPIVersionToInt(string networkAPIVersion)
+
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "Set to '*' to allow a new record set to be created, but to prevent updating an existing record set. Other values will result in error from server as they are not supported.")]
```
I understand, but I do not want to take the liberty to re-word how the feature team described their feature

grizzlytheodore replied:
improved

grizzlytheodore commented on src/Compute/Compute/Custom/VirtualMachineScaleSetsOperationsExtensions.cs:
```diff
@@ -13,12 +13,16 @@ public static partial class VirtualMachineScaleSetsOperationsExtensions
+            var ifMatchResult = virtualMachineScaleSet.GetIfMatchIfNoneMatch();
+            var ifMatch = ifMatchResult.Item1;
+            var ifNoneMatch = ifMatchResult.Item2;
             if (auxAuthHeader == null)
```
i was able to step over the code with the values being null. Compatible with IfMatch/IfNoneMatch being provided

Sandido commented:
@grizzlytheodore , think re-recording some other vmss tests as a sanity check would be worth it to show that they still work even with this feature? More than just a manual test. Unsure if this would be worth it.

grizzlytheodore commented:
> @grizzlytheodore , think re-recording some other vmss tests as a sanity check would be worth it to show that they still work even with this feature? More than just a manual test. Unsure if this would be worth it.

I was able to run 5 vmss tests that were added in the last 6 months successfully to check for regression

YanaXu commented:
Hi @grizzlytheodore, please resolve the conflict.

Sandido commented on src/Compute/Compute/Generated/VirtualMachineScaleSet/VirtualMachineScaleSetCreateOrUpdateMethod.cs:
```diff
@@ -401,5 +401,16 @@ private int convertAPIVersionToInt(string networkAPIVersion)
+
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "used to make a request conditional for the PUT and other non-safe methods. The server will only return the requested resources if the resource matches one of the listed ETag values. Omit this value to always overwrite the current resource. Specify the last-seen ETag value to prevent accidentally overwriting concurrent changes.")]
```
capital U

Sandido commented on src/Compute/Compute/Generated/VirtualMachineScaleSet/VirtualMachineScaleSetUpdateMethod.cs:
```diff
@@ -409,6 +409,18 @@ public override void ExecuteCmdlet()
+        [Parameter(
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "used to make a request conditional for the PUT and other non-safe methods. The server will only return the requested resources if the resource matches one of the listed ETag values. Omit this value to always overwrite the current resource. Specify the last-seen ETag value to prevent accidentally overwriting concurrent changes.")]
```
capital U

Sandido commented on src/Compute/Compute/VirtualMachine/Operation/NewAzureVMCommand.cs:
```diff
@@ -442,6 +442,18 @@ public class NewAzureVMCommand : VirtualMachineBaseCmdlet
+        [Parameter(
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "used to make a request conditional for the PUT and other non-safe methods. The server will only return the requested resources if the resource matches one of the listed ETag values. Omit this value to always overwrite the current resource. Specify the last-seen ETag value to prevent accidentally overwriting concurrent changes.")]
```
capital U

Sandido commented on src/Compute/Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs:
```diff
@@ -182,6 +182,18 @@ public class UpdateAzureVMCommand : VirtualMachineBaseCmdlet
+        [Parameter(
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "used to make a request conditional for the PUT and other non-safe methods. The server will only return the requested resources if the resource matches one of the listed ETag values. Omit this value to always overwrite the current resource. Specify the last-seen ETag value to prevent accidentally overwriting concurrent changes.")]
```
capital U

Sandido reviewed (approved):
Looks fine, just a spelling error and I think it's worth adding 1 example in one of the help files on how to use these new parameters.

YanaXu commented on src/Compute/Compute/help/New-AzVM.md:
````diff
@@ -738,6 +739,36 @@ Accept pipeline input: False
 Accept wildcard characters: False
 ```
 
+### -IfMatch
````
I'll merge this PR now. Feel free to create a new one to add more examples and fix the spelling error. Thanks.
